src/IMG_libpng.c

```diff
@@ -0,0 +1,2181 @@
+    if (!surface || !dst) {
+        SDL_SetError("Surface or SDL_IOStream is NULL");
+        return false;
+    }
+
+    if (!IMG_InitPNG()) {
+        return false;
+    }
+
+    struct png_op_vars vars;
+    bool result = false;
+
+    SDL_zero(vars);
+
+    result = LIBPNG_SavePNG_IO_Internal(&vars, surface, dst);
+
+    if (vars.png_ptr) {
+        lib.png_destroy_write_struct(&vars.png_ptr, &vars.info_ptr);
+    }
+    if (vars.color_ptr) {
+        SDL_free(vars.color_ptr);
+    }
+    if (vars.row_pointers) {
+        SDL_free(vars.row_pointers);
+    }
+    if (vars.source_surface_for_save && vars.source_surface_for_save != surface) {
+        SDL_DestroySurface(vars.source_surface_for_save);
+    }
+
+    if (!result && vars.error) {
+        SDL_SetError("%s", vars.error);
+    }
+
+    if (closeio && dst) {
+        SDL_CloseIO(dst);
+    }
+
+    return result;
+#endif
+}
+
+bool IMG_SavePNG(SDL_Surface *surface, const char *file)
+{
+    if (!surface || !file) {
+        SDL_SetError("Surface or file name is NULL");
+        return false;
+    }
+
+    SDL_IOStream *dst = SDL_IOFromFile(file, "wb");
+    if (!dst) {
+        SDL_SetError("Failed to open file %s for writing: %s", file, SDL_GetError());
+        return false;
+    }
+
+    return IMG_SavePNG_IO(surface, dst, true);
+}
+
+typedef struct
+{
+    png_uint_32 num_frames;
+    png_uint_32 num_plays;
+} apng_acTL_chunk;
+
+typedef struct
+{
+    png_uint_32 sequence_number;
+    png_uint_32 width;
+    png_uint_32 height;
+    png_uint_32 x_offset;
+    png_uint_32 y_offset;
+    png_uint_16 delay_num;
+    png_uint_16 delay_den;
+    png_byte dispose_op;
+    png_byte blend_op;
+    png_bytep raw_idat_data;
+    png_size_t raw_idat_size;
+} apng_fcTL_chunk;
+
+typedef struct
+{
+    SDL_IOStream *stream;
+    apng_acTL_chunk actl;
+    apng_fcTL_chunk *fctl_frames;
+    int fctl_count;
+    int fctl_capacity;
+    bool is_apng;
+    SDL_Surface *canvas;
+    SDL_Surface *prev_canvas_copy;
+    png_uint_32 current_idat_sequence;
+} apng_read_context;
+
+typedef struct
+{
+    SDL_IOStream *mem_stream;
+    SDL_IOStream *read_stream;
+    png_structp png_ptr;
+    png_infop info_ptr;
+    SDL_Surface *surface;
+    png_bytep *row_pointers;
+} DecompressionState;
+
+static SDL_Surface *decompress_png_frame_data(png_bytep compressed_data, png_size_t compressed_size,
+                                              int width, int height, int png_color_type, int bit_depth)
+{
+    /*
+     * Usually you'd directly decompress zlib but then we have to do defiltering and deinterlacing ourselves.
+     * We can decompress zlib then pass those jobs to libpng but then libpng expects a fully compressed data,
+     * therefore, we manually add chunks for header and other parts only for this given compressed data,
+     * tricking libpng to believe this is a normal PNG file, then make it defilter (if any) and deinterlace
+     * (if any) for us.
+     */
+
+    DecompressionState state;
```

Comment:
Similarly the compiler would be allowed to put some or all of the fields of this struct in registers, making them susceptible to being clobbered (overwritten with garbage values) by `longjmp()`.